src/chart/layout/mod.rs

```diff
@@ -0,0 +1,995 @@
+        // not spilling over off the drawing_area.
+        let left_spill = top_tick_labels_extent.x0.min(bottom_tick_labels_extent.x0);
+        if left_spill < 0 {
+            let (w, h) = self
+                .nodes
+                .get_left_tick_label_size()
+                .ok_or_else(|| LayoutError::ExtentsError)?;
+
+            self.nodes
+                .set_left_tick_label_size((w - left_spill) as u32, h as u32)?;
+        }
+        let right_spill =
+            canvas_w as i32 - top_tick_labels_extent.x1.max(bottom_tick_labels_extent.x1);
+        if right_spill < 0 {
+            let (w, h) = self
+                .nodes
+                .get_right_tick_label_size()
+                .ok_or_else(|| LayoutError::ExtentsError)?;
+
+            self.nodes
+                .set_right_tick_label_size((w - right_spill) as u32, h as u32)?;
+        }
+
+        // Layouts are cached, so if we didn't change anything, this is a very cheap function call
+        self.nodes.layout(canvas_w as u32, canvas_h as u32)?;
+
+        // It may be the case that parts of the label text "spill over" into the margin
+        // to the top/bottom of the chart_area. We want to make sure that we're
+        // not spilling over off the drawing_area.
+        let top_spill = left_tick_labels_extent.y0.min(right_tick_labels_extent.y0);
+        if top_spill < 0 {
+            let (w, h) = self
+                .nodes
+                .get_top_tick_label_size()
+                .ok_or_else(|| LayoutError::ExtentsError)?;
+
+            // When the left/right tick label extents were computed, the bottom/top tick labels
+            // had zero size. We only want to increase their size if needed. Otherwise, we should
+            // leave them the size they are.
+            self.nodes
+                .set_top_tick_label_size(w as u32, h.max(-top_spill) as u32)?;
+        }
+        let bottom_spill =
+            canvas_h as i32 - left_tick_labels_extent.y1.max(right_tick_labels_extent.y1);
+        if bottom_spill < 0 {
+            let (w, h) = self
+                .nodes
+                .get_bottom_tick_label_size()
+                .ok_or_else(|| LayoutError::ExtentsError)?;
+
+            // When the left/right tick label extents were computed, the bottom/top tick labels
+            // had zero size. We only want to increase their size if needed. Otherwise, we should
+            // leave them the size they are.
+            self.nodes
+                .set_bottom_tick_label_size(w as u32, h.max(-bottom_spill) as u32)?;
+        }
+
+        // Layouts are cached, so if we didn't change anything, this is a very cheap function call
+        self.nodes.layout(canvas_w as u32, canvas_h as u32)?;
+
+        Ok(())
+    }
+
+    /// Helper function for drawing ticks. This function will call
+    /// `draw_func(pixel_coords, tick, axis_side)` for every tick on every axis.
+    fn draw_ticks_helper<F>(
```

Comment:
If my understand  was correct, currently the tick is drawn by newly added code instead of the code in ChartContext?

Reply:
That's correct. If I could turn an `AsRangedCoord` into an iterator that iterated through all the tick marks and labels, I could use the existing data structures to do it (this is what I couldn't figure out how to do).

I need to find the extents of each text item/tick mark to compute the correct amount of space.

Reply:
Hmm, but I think we probably don't want having two different set of drawing code. If this is the case, I may need to think about unify those code into one. Also, I have the feel that this change seems a little bit big. So I think it's better to split it into parts.

Apologize for this, but I feel we should split the change into parts, since I think we are currently on a way leading to significant refactoring. 

If this is the case, I am going to dig into commit 9938d7b, tweak a little bit and merge it.  I will leave this PR open. After that let's figure out some better way to integrate rest part of the PR properly. 

Please let me know what's your idea.

Reply:
That sounds good. I can rebase from there after you've tweaked it.